support/oidc-discovery-provider/domain_policy.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"fmt"
+
+	"golang.org/x/net/idna"
+)
+
+type DomainPolicy = func(domain string) error
+
+// DomainAllowlist returns a policy that allows any domain in the given domains
+func DomainAllowlist(domains ...string) (DomainPolicy, error) {
+	allowlist := make(map[string]struct{}, len(domains))
+	for _, domain := range domains {
+		domainKey, err := toDomainKey(domain)
+		if err != nil {
+			return nil, err
+		}
+		allowlist[domainKey] = struct{}{}
+	}
+	return func(domain string) error {
+		domainKey, err := toDomainKey(domain)
+		if err != nil {
+			return err
+		}
+		if _, allowed := allowlist[domainKey]; !allowed {
+			return fmt.Errorf("domain %q is not allowed", domain)
+		}
+		return nil
+	}, nil
+}
+
+// AllowAnyDomain returns a policy that allows any domain
+func AllowAnyDomain() DomainPolicy {
+	return func(domain string) error {
+		_, err := toDomainKey(domain)
+		return err
+	}
+}
+
+func toDomainKey(domain string) (string, error) {
+	domainKey, err := idna.Lookup.ToASCII(domain)
```

Comment:
The intent here is to .... allow the provider to be configured with domain names that have non-ascii characters and then do the punycode translation here under the covers?

Would it be better to make the users provide the punycode explicitly? Many other places in the codebase are moving towards literal expression of things like this after CVE-2021-27099. Just a thought...

Reply:
That's a good idea. I'll require the domain to already be expressed in punycode.